### PR TITLE
add sw_64 support

### PR DESCRIPTION
**Pull request by weigang-li**
add sw_64 support

### CODE REVIEW

File: src/atomic_ops_stack.c [hunk 1]
```diff
@@ -81,7 +81,7 @@ AO_API AO_uintptr_t *AO_stack_next_ptr(AO_uintptr_t next)
     } /* extern "C" */
 # endif
 
-# if defined(__alpha__) && (__GNUC__ == 4)
+# if (defined(__alpha__) || define(__sw_64__)) && (__GNUC__ == 4)
```
Comment (ivmai) [inline src/atomic_ops_stack.c:84]: define -> defined

Comment (ivmai) [reply src/atomic_ops_stack.c:84]: Hmm. Do we really need this workaround intended for old gcc version on alpha?

File: src/Makefile.am [hunk 2]
```diff
@@ -85,6 +85,7 @@ nobase_private_HEADERS = atomic_ops/ao_version.h \
         \
           atomic_ops/sysdeps/gcc/aarch64.h \
           atomic_ops/sysdeps/gcc/alpha.h \
+          atomic_ops/sysdeps/gcc/sw_64.h \
```
Comment (ivmai) [inline src/Makefile.am:88]: order item lexicographically

File: src/atomic_ops/sysdeps/gcc/sw_64.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+AO_nop_full(void)
+{
+  __asm__ __volatile__("memb" : : : "memory");
+}
+#define AO_HAVE_nop_full
+
+AO_INLINE void
+AO_nop_write(void)
+{
+  __asm__ __volatile__("memb" : : : "memory");
```
Comment (ivmai) [inline src/atomic_ops/sysdeps/gcc/sw_64.h:36]: I don't see how the implementation differs from AO_nop_full - if so, then just drop AO_nop_write

File: src/atomic_ops/sysdeps/gcc/sw_64.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (ivmai) [inline src/atomic_ops/sysdeps/gcc/sw_64.h:1]: I see that you've copied and edited gcc/alpha.h file. Is sw_64 very close to Alpha architecture regarding memory consistency? Why not just include "generic.h" and that's all? (see e.g. gcc/e2k.h)

**Review by ivmai — CHANGES_REQUESTED** [review]
Please study gcc/e2k.h - it merely includes gcc/generic.h and that's all!

File: src/atomic_ops/sysdeps/gcc/sw_64.h (hunk 4 above)
Comment (weigang-li) [reply src/atomic_ops/sysdeps/gcc/sw_64.h:1]: sw_64 and alpha architecture are not the same. Therefore it needs to be modified like this

File: src/atomic_ops/sysdeps/gcc/sw_64.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Permission is hereby granted to use or copy this program
+ * for any purpose,  provided the above notices are retained on all copies.
+ * Permission to modify the code and to distribute modified code is granted,
+ * provided the above notices are retained, and a notice that the code was
+ * modified is included with the above copyright notice.
+ *
+ */
+
+#include "../loadstore/atomic_load.h"
```
Comment (ivmai) [inline src/atomic_ops/sysdeps/gcc/sw_64.h:18]: please remove entire content of this file and add just one line:
#include "generic.h"
Will it work for you as expected?

**Review by ivmai — CHANGES_REQUESTED** [review]

**Comment by ivmai** [comment]
Okay, does it work for you as expected?

**Comment by weigang-li** [comment]
> 好的，它对您有预期的效果吗？

Yes, it can achieve my expected effect

**Comment by ivmai** [comment]
> Yes, it can achieve my expected effect

Please drop the patch and check the unmodified libatomic_ops - it should still work for you, because sw_64 platform was already supported.

File: src/Makefile.am [hunk 6]
```diff
@@ -100,6 +100,7 @@ nobase_private_HEADERS = atomic_ops/ao_version.h \
           atomic_ops/sysdeps/gcc/s390.h \
           atomic_ops/sysdeps/gcc/sh.h \
           atomic_ops/sysdeps/gcc/sparc.h \
+          atomic_ops/sysdeps/gcc/sw_64.h \
```
Comment (ivmai) [inline src/Makefile.am:103]: CMakeLists.txt file should be modified as well.

**Comment by ivmai** [comment]
Currently the patch does nothing. I'm dropping it.

**Pull request closed without merging**